current/en-us/5. templating/4. custom-elements.md

```diff
@@ -378,3 +378,405 @@ There are lots of options that allow you to change how custom elements work. The
+    const content = document.createElement("div");
+    content.setAttribute("show.bind", `activeTabId=='${i}'`);
+    content.append(...Array.from(tab.childNodes));
+    contentTemplate.content.appendChild(content);
+
+    node.removeChild(tab);
+  }
+
+  // Activate the first tab
+  const bindingEngine = Container.instance.get(BindingEngine);
+  instruction.attributes = {
+    ...instruction.attributes,
+    "active-tab-id": bindingEngine.createBindingExpression("activeTabId", "'0'")
+  };
+
+  node.append(headerTemplate, contentTemplate);
+  return true;
+}
+
+@processContent(processTabs)
+export class Tabs {
+  @bindable activeTabId;
+  showTab(tabId) {
+    this.activeTabId = tabId;
+  }
+}
+```
+
+```TypeScript Tabs.ts [variant]
+import { BehaviorInstruction, bindable, BindingEngine, Container, processContent } from "aurelia-framework";
+
+@processContent(Tabs.processTabs)
+export class Tabs {
+  // A static method in the custom element class itself is more compact than an external function as shown the JavaScript version.
+  // But both serves same purpose.
+  public static processTabs(_compiler, _resources, node: HTMLElement, instruction: BehaviorInstruction) {
+    // first create 2 templates for the replaceable parts
+    const headerTemplate = document.createElement("template");
+    headerTemplate.setAttribute("replace-part", "header");
+    const contentTemplate = document.createElement("template");
+    contentTemplate.setAttribute("replace-part", "content");
+
+    // process all tabs
+    const tabs = Array.from(node.querySelectorAll("tab"));
+    for (let i = 0; i < tabs.length; i++) {
+      const tab = tabs[i];
+
+      // add header
+      const header = document.createElement("button");
+      header.setAttribute("click.delegate", `showTab('${i}')`);
+      header.innerText = tab.getAttribute("header");
+      headerTemplate.content.appendChild(header);
+
+      // add content
+      const content = document.createElement("div");
+      content.setAttribute("show.bind", `activeTabId=='${i}'`);
+      content.append(...Array.from(tab.childNodes));
+      contentTemplate.content.appendChild(content);
+
+      node.removeChild(tab);
+    }
+
+    // Activate the first tab
+    const bindingEngine: BindingEngine = Container.instance.get(BindingEngine);
+    instruction.attributes = {
+      ...instruction.attributes,
+      "active-tab-id": bindingEngine.createBindingExpression("activeTabId", "'0'")
+    };
+
+    node.append(headerTemplate, contentTemplate);
+    return true;
+  }
+
+  @bindable public activeTabId: string;
+  public showTab(tabId: string) { this.activeTabId = tabId; }
+}
+```
+
+The `processTabs` method used in `@processContent` may look overwhelming, but all it does is mostly pure DOM manipulation. The function takes four arguments, out of which `node` holds the DOM fragment for `<tabs>...</tabs>` used in `app.html`. And then in this function, we transforms the original DOM fragment to something like below.
+
```

Comment:
Should there be a before/after comparison for the "look" of `<tabs>...</tabs>` as the content are a bit far away from each other?

Reply:
Do you mean like a side by side two column comparison? If yes, then I am not really sure how to do that in pure markdown. That might be possible if I use HTML, but then again follow the question of code highlighting and all. Or just duplicating the previous code fragment right before this one, will do? Or have I completely misunderstood?